README.md

```diff
@@ -1105,6 +1110,59 @@ However, the following rules of thumb are useful to keep in mind:
   Alternatively, the entire OrientDB class and all edges that may point to it may be omitted
   entirely from the GraphQL schema.
 
+## Query cost estimation
+
+Due to GraphQL's easy-to-use nature, queries that use a lot of memory and/or have a long execution
+time can happen often. For better user-friendliness, it's worthwhile to estimate a query's cost (in
+terms of time taken and memory used) using statistics without executing the query on the database.
+Since the time and memory required for query execution varies depending on the underlying database,
+as a rough proxy, the cost estimator estimates the number of result rows for the given query i.e.
+the query's cardinality.
+
+These estimates can be used to improve user-friendliness for query execution.
+For example, the estimates can be used to warn users about a query that asks for too many results.
+
+### Usage of query cost estimation
+
+To get estimates for the number of result rows for a given query and statistics object, use the
+`estimate_query_result_cardinality()` function in
+`graphql_compiler/cost_estimation/cardinality_estimator.py`.
+
+### Statistics
+
+Before estimates can be generated, you need to provide statistics about the graph data. The
+applicable classes are in `graphql_compiler/cost_estimation/statistics.py`. Currently, two classes
+are implemented:
+- `Statistics`, an abstract base class that describes the interface the cost estimator expects for
+  statistics. This allows users to create their own implementations for the statistic class.
+- `LocalStatistics`, an implementation of `Statistics`, where the statistics are provided during
+  initialization, and can't be modified thereafter.
+
+Currently, three statistics are supported. `get_class_count()` must be provided for cardinality
+estimation. The other two are optional, as they can be roughly approximated.
```

Comment:
> they can be roughly approximated.

Worthwhile discussing how they can be approximated? Up to you.

Reply:
I don't think this should be in the README. There are many reasons: the estimator's implementation is only prototypical; the current implementation is well-documented, but it's not really simple; and most important of all: the user shouldn't need to care how it's implemented.

Reply:
I disagree. If the user does not have an intuition for how it works, he or she will not trust it.

Reply:
In the readme, at no point do we describe how graphql compilation works (GraphQL->AST->IR->database language). I don't see why a user should understand how cost estimation works and how the IR is used, when they don't even know what IR is.

I've noticed we disagree on this. I'll speak with other members of the team, and see what's the best course of action.

Reply:
How about describing details like these in the comments, and in the README tell the user to read the comments in X file if they want to know more details about how the statistics are calculated? This way we don't put too much stuff in the README but it's still clear exactly where they can get an intuition of how things work.